[0012] Add node discovery protocol (#33)
nervosnetwork/rfcs: rfcs/0008-node-discovery/0008-node-discovery.md

---

> Number: 8

Reviewer: `8` has been taken

Category: Standards Track
Status: Draft

> Author: Linfeng Qian / JinYang Jiang

Reviewer: Should use comma `,` to separate authors.

Organization: Nervos Foundation
Created: 2018-11-28
---

# CKB Node Discovery Protocol

CKB node discovery protocol mainly refers to [Satoshi Client Node Discovery][0]. The differences between them are summarized below:

* The node version number is included in the `GetNodes` message.
* The `Nodes` message is used to periodically broadcast all nodes currently connected.

> * We use `multiaddr` as the format of node addresses  (There is no `/p2p/` field here and if it's considered as *misbehavior*, a low score will be given.)

Reviewer: ```suggestion
* We use `multiaddr` as the format of node addresses  (It MUST NOT include `/p2p/` segment otherwise it's considered as *misbehavior* and a low score will be given.)
```

Every time client startup, if PeerStore's address list is empty, it will try to issue DNS requests to initialize address list, if DNS requests failed it will fallback to the hard-coded address list.

> ## Discovery Methods

Author: `Discovery Methods` is more nature for me ...

### DNS Addresses

> When first time startup (bootstrap stage), if discovery service is needed, local node then issues DNS requests to learn about the addresses of other peer nodes. The client includes a list of host names for DNS services that are seeded.

Reviewer: Can we add a definition for `seeded`, i.e. what kind of nodes can be called `seeded`?

Author: Added

> When first time startup (bootstrap stage), if discovery service is needed, local node then issues DNS requests to learn about the addresses of other peer nodes. The client includes a list of host names for DNS services that are seeded. DNS server addresses can be replaced by command line arguments.

Reviewer: ```suggestion
At the first time startup (bootstrap stage), if the discovery service is needed, the local node will issue DNS requests to learn about the addresses of other peer nodes. The client includes a list of seed hostnames for DNS services.
```

Reviewer: Removed `DNS server addresses can be replaced by command line arguments.` which is an implementation detail, not protocol.

### Hard-Coded "Seed" Addresses

> The client contains hard-coded IP addresses that represent CKB nodes. Those addresses are used only if all DNS requests failed. Once the local node has enough addresses (presumably learned from the seed nodes), the client will close seed node connections to avoid overloading those nodes.

Reviewer: ```suggestion
The client contains hard-coded IP addresses that represent CKB nodes. Those addresses are used only if all DNS requests failed. Once the local node has enough addresses (presumably learned from the seed nodes), the client SHOULD close seed node connections to avoid overloading those nodes.
```

> ### Protocol Message

Author: It's generic reference here, no need plural form?

> The client contains hard-coded IP addresses that represent CKB nodes. Those addresses only are used only if all DNS requests failed. Once the local node has enough addresses (presumably learned from the seed nodes), the client will close seed node connections to avoid overloading those nodes. "Seed" addresses can be specified manually with command-line arguments.

Reviewer: There're two 'only' in the 2nd sentence.

#### `GetNodes` Message
When all the following conditions are met, the local node will send a `GetNodes` message:

  1. It's an outbound connection (for resisting [fingerprinting attack][3]).
  2. The other node's version must bigger than a preset value.
  3. The number of addresses currently stored is less than `ADDRESSES_THRESHOLD` (default 1000). 

#### `Nodes` Message

> When the client receives a `GetNodes` request, it will return a `Nodes` message if this kind of reception is the first time and the connection is an inbound connection, the `announce` field is set to `false`. At regular intervals, local node will broadcast all connected `Node` information in `Nodes` message to all connected nodes, the `announce` field is set to `true`. When local node received a `Nodes` message and it's `announce` field is `true`, local node will relay those node addresses that are [routable][1].

Reviewer: `it will ...` -> `it SHOUILD ...` or `it MUST ...`?

Reviewer: We'd better review all `will` in this document.

Author: Yeah

> Here `announce` field is to distinguish a `Nodes` as a response of `GetNodes` or broadcast message, so it's convenient to apply different rules for punishing misbehaviors. The main rules:

Author: I think `convenient` is more accurate here. Because `announce` is just a helper field, not a required field.

* A node can only send one `Nodes` message (announce=false) as a response of `GetNodes` message.

> * A node's broadcast messages only the first `Nodes` message (announce=true) can include more than `ANNOUNCE_THRESHOLD` (default 10) node information, in case other peer send reduplicative `Node` message.

Reviewer: ```suggestion
* Among a node's broadcast messages only the first `Nodes` message (announce=true) can include more than `ANNOUNCE_THRESHOLD` (default 10) node information, in case other peer send reduplicative `Node` message.
```

And what does 'reduplicative' mean?

Author: repeated?

> * A node's broadcast messages only the first `Nodes` message (announce=true) can include more than `ANNOUNCE_THRESHOLD` (default 10) node information, in case other peer send reduplicative `Node` message.

Reviewer 2: ```suggestion
 * Only the first `Nodes` message (announce=true) of broadcast messages in a node can include more than `ANNOUNCE_THRESHOLD` (default 10) node information. This can prevent other peers from sending reduplicative `Node` message.
```

Author: I may follow jan's suggestion here (see above) ~

The number of `addresses` field of each `Node` in all `Nodes` messages cannot exceed `MAX_NODE_ADDRESSES` (default 3).

> 

Reviewer: Some examples of `GetNodes` and `Nodes` messages will help to understand.

Author: I think it's obvious when you see the diagrams below.

## Resist Typical Attacks
### Fingerprinting Attack
[Related paper][3]

`GetNodes` can only send to an outbound connection.

## Flow Diagram
### Node Bootstrap
![](images/bootstrap.png)
### Send `GetNodes` Message
![](images/get-nodes.png)
### Announce Connected Nodes
![](images/announce-nodes.png)

> ## Data Structures

Reviewer: I suggest moving data structures before flow charts.

Author: ok

---
Number: 8

> Category: Informational

Reviewer 3: ```suggestion
Category: Standards Track
```

### Protocol Message
#### `GetNodes` Message
When all the following conditions are met, the local node will send a `GetNodes` message:

  1. It's an outbound connection (for resisting [fingerprinting attack][3]).
  2. The other node's version must bigger than a preset value.
  3. The number of addresses currently stored is less than 1000. 

#### `Nodes` Message
When the client receives a `GetNodes` request, it will return a `Nodes` message if this kind of reception is the first time and the connection is an inbound connection, the `announce` field is set to `false`. At regular intervals, local node will broadcast all connected `Node` informations in `Nodes` message to all connected nodes, the `announce` is field is set to `true`. When local node received a `Nodes` message and it's `announce` field is `true`, local node will relay those node addresses that are [routable][1] .

Here `announce` field is for distinguish a `Nodes` as an response of `GetNodes` or broadcast message, so it's convient to apply different rules for punishing misbehaviors. The main rules:

> * A node can only send one `Nodes` message (announce=true) as an response of `GetNodes` message.

Reviewer 4: The field `announce` in the message `GetNodes` response `Nodes` is false. There must be typo，the same as Chinese version

> * A node's broadcast messages only the first `Nodes` message can include more than 10 node informations.

Reviewer 5: Why? please explain the intention of this behavior
